osbrain/tests/test_proxy.py

```diff
@@ -119,6 +122,51 @@ def ping(self):
         InitTimeoutProxy('foo', timeout=1.)
 
 
+@pytest.mark.parametrize('timeout', [-1, 1])
+def test_agent_proxy_wait_running(nsproxy, timeout):
+    """
+    Using `wait_for_running` on a proxy after initialization should block until
+    the agent is running or time out.
+    """
+    AgentProcess('agent').start()
+
+    # Get "offline" proxy
+    agent = Proxy('agent')
+    time0 = time.time()
+    Timer(abs(timeout)/2, agent.run).start()
+
+    proxy = Proxy('agent').wait_for_running(timeout=timeout)
+    assert proxy.ping() == 'pong'
+    assert abs(timeout)/2 <= time.time() - time0 <= abs(timeout)
+
+
+def test_agent_proxy_wait_running_0_seconods(nsproxy):
+    """
+    Using `wait_for_running` on a proxy after initialization should block until
+    the agent is running or time out.
+    """
+    run_agent('agent')
+
+    proxy = Proxy('agent').wait_for_running(timeout=0)
+    assert proxy.ping() == 'pong'
+
+
+@pytest.mark.parametrize('timeout', [0, 1])
+def test_agent_proxy_wait_running_timeout(nsproxy, timeout):
+    """
+    Check that the `wait_for_running` method times out if the agent is not
+    running after the specified number of seconds.
+    """
+    AgentProcess('agent').start()
+
+    time0 = time.time()
+    with pytest.raises(TimeoutError) as error:
+        Proxy('agent').wait_for_running(timeout=timeout)
+
+    assert "Timed out" in str(error.value)
+    assert time.time() - time0 >= timeout
```

Comment:
It would be great if you could add another assertion for the upper bound. Right now the TimeoutError could be raised after 3 seconds for both tests, for example. You are just checking the assertion occurs after the timeout, but not when.

Maybe something like: `assert time.time() - time0 < timeout + .5`.

Reply:
Not very critical, but OK :ok_hand: